### ISSUE TITLE
qBittorrent import tool

### ISSUE THREAD

**Issue by slrslr** [BiglySoftware/BiglyBT#1058]
* BiglyBT Version Number: 1.9

Hello, it can be handy if there is tool/function that will import torrents from qbittorrent (torrent client).

**Next 7 posts does not solve this issue and can be ignored.**

**Comment by PrototypeZero** [comment]
Whats wrong with adding the torrent to Bigly and just pointing the destination to the files already downloaded with qbittorent?  Bigly will automatically check the files and download what it needs to complete the torrent (no need to start again).

**Comment by slrslr** [comment]
@PrototypeZero if you mean [these steps](https://github.com/BiglySoftware/BiglyBT/issues/17#issuecomment-437924531), i tried them and commented in that issue & not worked. If you mean different steps, i would appreciate if you write them here.

Note that i have more than thousand torrents acros many many directories, some have common files and some have payload files and folders with custom names.

**Comment by PrototypeZero** [comment]
Well how I do it is 

Make sure Options->Files-> When opening a torrent show dialog is set to always.
Find/download the torrent file and open it, change save location to the base folder of the torrent.
Press OK.
Depending on if your using the beta or not you will get a dialog saying it found files (last few betas dont show this for me atleast).
Let it continue and it will check and resume downloading only the parts you need.

**Comment by slrslr** [comment]
Thanks, but sorry, i am looking for the import function. Not to manually find and select path for 1000+ torrents.

**Comment by PrototypeZero** [comment]
If you do have all the torrent files then...

What your probably looking for then is to use the automatic torrent import folder with all the torrent files in it, and make sure the data files are in a default directory, that will work (I've done to import about 100 torrents at once in the past).  It will perform as I previously said checking the torrents and only downloading what it needs to.  

This should work, although as said 1,000's of active torrents is probably pushing any single BT client including BiglyBT.

**Comment by slrslr** [comment]
No, i mentioned i am using many directories. and custom ones + custom payload file names, so i need import tool.

**Comment by PrototypeZero** [comment]
Wouldn't take too long to do as I have suggested along with a little work, but I guess you don't want to do any work yourself.  I doubt the devs would add this just for your unusual case.

**Comment by as-muncher** [comment]
> Wouldn't take too long to do as I have suggested along with a little work, but I guess you don't want to do any work yourself. I doubt the devs would add this just for your unusual case.

I just recently tried this, having BiglyBT look for existing data files.  The problem is that it as of 2.6 will only add the files with the same extension.  It doesn't look at the file with a .!qB extension, and see how much of it is salvageable.  It means currently that I'd have to delete all those extensions for those files.  Another step that possibly could be avoided easily if BiglyBT just automatically found files or recognised files with any extension, or at least files with the filename the same but not the extension, and see if it could integrate those files, at least that would be helpful.

**Comment by as-muncher** [comment]
Then at least, if BiglyBT would recognise whatever file extensions, I wouldn't have to manually use a renamer program to delete the extension.  Right now, I use Rename Master.  And even though I have BiglyBT set up to append the .!qB extension to incomplete files, it still doesn't recognise files with that extension.  So, hopefully, if BiglyBT would recognise at least the filename without the extension and try to salvage what it can from that file for a matching file in the .torrent file, then that would save the effort of having to rename the file and then have BiglyBT search it again.

**Comment by ferdnyc** [comment]
Having SFEDF (or BiglyBT in general) ignore _all_ file extensions strikes me as dangerous, as it makes the torrent contents ambiguous. Say there's a torrent that includes PGP signature files (`.asc`) files) for each of its members. And now say you scan a directory like this:

```text
somepkg.tgz.part
somepkg.tgz.asc.part
somepkg.tgz.txt
somepkg.tgz.txt.asc
otherpkg.tgz
otherpkg.tgz.asc
otherpkg.tgz.txt.part
otherpkg.tgz.txt.asc.part
```

...Should `somepkg.tgz.txt.asc` be checked as possibly being the file `somepkg.tgz.txt.asc`, or `somepkg.tgz.txt`, or `somepkg.tgz`? Yeah, you can try to disambiguate with file sizes, but many smaller files (and all `.asc` files) will be the same size.

Better, I think, to have a list of extensions to strip off configurable in the SFEDF dialog. E.g:

**Ignore these extensions when scanning files:** `.part,.!qB`

**Comment by ferdnyc** [comment]
> **Ignore these extensions when scanning files:** `.part,.!qB`

On second thought, that label makes it sound too much like "Ignore _files with_ these extensions..." which is a different thing entirely. So perhaps:

**Strip these extensions when matching filenames:** `.part,.!qB`

**Comment by ferdnyc** [comment]
On the self-serve front, the command-line [`mmv` tool](https://github.com/rrthomas/mmv) is one option for quickly stripping away file extensions in a set of directories.

This command, for example, will strip the `.!qB` extension off any file, at any depth, under the `/start/directory`:

```bash
find /start/directory -type f -execdir \
  mmv -v '*.!qB' '#1' \;
```

If you want to play it safer and reduce the work done, you can replace `-type f` with `-name '*.!qB'`, and it'll only process directories containing at least one matching filename.

Windows users, `¯\_(ツ)_/¯`

**Comment by as-muncher** [comment]
@ferdnyc  So far, BiglyBT is working okay as is, if I choose to have the incomplete file extension as .!qB, with working with files that currently have .!qB from qbittorrent.  I think just some minor tweaks have to be made, and I have raised a few Github issues for that.  I don't know how many people save their .torrent files.  I do and like to save the data and the .torrent file together, which maybe is not great practice, because I should probably have one other copy somewhere saved of the .torrent file.  Right now, if I open the .torrent file in BiglyBT, it searches through the data, and if the data has a .!qB extension and I have the option in BiglyBT to save incomplete files with the .!qB extension, then it works out okay, I think always if not most of the time.  Perhaps if something were automated, it could find the .torrent files that qbittorrent is using, especially if they're saved as <infohash>.torrent and import that into BiglyBT, and then something automated to add the .!qB extension to incomplete files, in the settings of BiglyBT.   But as one user commented that he wanted a tool with changing multiple file names etc., in multiple directories, it might make sense to have something look into where qbittorrent saves each torrent, make it easy to switch from qbittorrent to BiglyBT.  Right now, I'm using BiglyBT mostly for swarm merging, because I find the downloading rate quite a bit restricted using BiglyBT vs qbittorrent.  And I want to use BiglyBT to focus on a few priority torrents.  

I personally want BiglyBT to search for files with .part and .!qB or whatever the extension.  If the data is the same, then great, I'll take it.  I don't think BiglyBT goes strictly on file size.

**Comment by Dramaman** [comment]
My assumption is that the .!qB files will import if you simply rename them to remove the extension?
If so, can't you use a mass file renaming application?
https://medevel.com/22-file-renamer-list/

Assuming qbittorrent can't simply rename the file names to remove the extension itself, Perhaps their developers can add that as a feature to their clients?

As long as you don't mess with the directory paths, or know what adjustments to make to accommodate any changes that you did in the other client, It should be pretty straight forward to import once you revert the name and directory path changes.

Users can also opt not to add those custom file extensions.

**Comment by as-muncher** [comment]
@Dramaman If you read the posts I made, you will see that the system currently works fine with the .!qB extension just as long as you tell BiglyBT to add the .!qB extension to incomplete files.

**Comment by NiKola-UE** [comment]
Here, let me get involved in this issue. 

QBittorrent is a really good FOSS program and the best of it should be implemented. 

If for some reason it is necessary to temporarily archive or compresse the files, the codes of 7-Zip, PeaZip and 7-Zip ZS can be used for this. 

In addition to IPFS for data processing and storage, SeaweedFS can definitely use, that is not so well known, but shows a lot of promise and is still being developed, which is very good. 

I recently opened 

[this issue, ](https://github.com/BiglySoftware/BiglyBT/issues/3205)



so here are some more good suggestions.

**Comment by NaPurHab** [comment]
As I am just coming into this, I have a suggestion.  I have an issue I will post seperately.  But I have BiglyBT and qBittorrent running at the same time, both through my VPN.  I added a torrent I have on BiglyBT to qBittorrent and it quickly downloaded it internally in my system.  Sure, I ended up with 2 copies. But the conversion works.

So to move from qBittorrent to BiglyBT.  
In BiglyBT under **Options -> Files -> Torrents**
Set an Import Directory. Or Several if you want to sort to TAGS.
Example _(Note these paths are on Google Drive.  Makes it easy to add torrent from anywhere. Just add to the correct Google Drive Path and BiglyBT picks them up.)_
C:\Users\<user>\My Drive\Torrents\Plex Movies
C:\Users\<user>\My Drive\Torrents\Plex TV            
then apply a TAG to each i.e.   Movies and TV

Then find the torrent files for qBittorrent on disk and copy to the new import directory(s). If you create multiple above, you can sort them as you place them
BiglyBT will import and then DL locally from qBitorrrent.

You will temporarily have 2 copies of each torrent. So make sure to have enough free space. But you will have no issues with the incomplete files in qBittorrent converting, they will just DL internally to BiglyBT.

It works a treat, G'luck.

**Comment by Ghost-chu** [comment]
I have upload it to Github, see: https://github.com/PBH-BTN/biglybt-qbittorrent-migrator

As I said, it only does the basics. biglyBT's internal code is so complex that it's hard for me to go any further.
Documentation on how to use it has been placed in README.md.

**Comment by NiKola-UE** [comment]
@Ghost-chu, 


Good job, really! 




But can this be made into a tool or plugin/addon (like the Tor, I2P, IPFS, etc.) so that it can be directly downloaded from the program menu, ie. integrated without the need to go to GitHub and download it as a file, and then manually add it to the program? I think this would be more practical.
